features/268-next-cf/api.md

```diff
@@ -0,0 +1,625 @@
+
+SPA requires client side routing, because all pages ares served by one single html file. Thus requests that routes to different path need to be responsed with that html file.
+
+With s3 or cloudfront, they both provide configuration for custom error response.
+- s3: `ErrorDocument` in `WebsiteConfiguration`, see https://docs.aws.amazon.com/en_us/AmazonS3/latest/API/RESTBucketPUTwebsite.html
+- cloudfront: `CustomErrorResponse`, see https://docs.aws.amazon.com/en_us/cloudfront/latest/APIReference/API_UpdateDistribution.html
+
+The following is a platform independent approach:
+```yaml
+static:
+  - src: build/index.html
+    path: /
+  - src: build/asset
+    path: /static
+```
+
+Following the route order, request path that matches `/static/*` would be served with the content in folder `build/asset` and the all other request would be served with `build/index.html`.
+
+For example, if someone enter url `myapp.skygear.io/page/1` in the browser, `build/index.html` will serve the request. In the html file, there will be a script element with src, for example, `https://myapp.skygear.io/static/bundle.js`. The js runtime would handle the path `/page/1` at client side and route to the corresponding page.
+
+# Route matching
+
+To easily distinguish gear route and non-gear route, I suggest we add the following two rules:
+- All gear routes start with `_`, e.g. auth gear routes are `/_auth/`, signup path would be `/_auth/signup`
+  - So any request with its path start with `_` would be considered request to gear
+- Non-gear path cannot be start with `_`
+
+Here is the route matching rules when a request enter the cluster:
+- If the path starts with `_`, it would try find the gear by matching `/_{gear_name}` first
+  - If a gear is found, the request would be forwarded to that gear
+  - Otherwise, return function not found error
+- Path of `http-handler` and http `function` are exact match
+- Path of `http-service` are matched with prefix
+- Longer path will be matched before the shorter ones
+- If none of the registered path is matched, return function not found error
+
+Static assets share another domain, so routing is done separately, the rules are similar:
+- Path are matched with prefix
+- Longer path will be matched before shorter ones
+- If none of the registered path is matched, return asset not found error
+
+For example, with the following skycli configuration:
+
+```yaml
+cf:
+  functionABC:
+    type: http-handler
+    path: /functionABC
+    # other config
+  function-server:
+    type: http-service
+    path: /function
+    # other config
+  api-server:
+    type: http-service
+    path: /api
+    # other config
+
+static:
+  - src: index.html
+    path: /
+  - src: asset
+    path: /static
+```
+
+- `https://myapp.api.skygear.io/function/ABC` -> `function-server`, forwarded path `/ABC`
+- `https://myapp.api.skygear.io/functionABC` -> `functionABC`
+- `https://myapp.api.skygear.io/functionABCD` -> `function-server`, forwarded path `ABCD`
```

Comment:
`https://myapp.api.skygear.io/functionABCD` -> `function-server`, forwarded path `ABCD`, this match is a little weird to me.  May I know the reason why it is a prefix match for `function-service`? Which implies `https://myapp.api.skygear.io/functionABCD` and `https://myapp.api.skygear.io/function/ABCD` got the same route.

Reply:
`http-service` is a prefix match instead of exact match because it allows the service to serve different resources by the http request path. Otherwise exact match won't allow the service to do any routing.

For the gateway,`https://myapp.api.skygear.io/functionABCD` and `https://myapp.api.skygear.io/function/ABCD` would route to the same service, but the service, by its routing, may be handled by different handlers.

Reply:
Any reason for prefix match for cloud functions vs longest match for services? Would it be bad if we simply use longest match for everything?

Reply:
I'm a bit confused with prefix match vs longest match, maybe they differ in handling `/`? Maybe its also what @kaojohnny was asking that i didn't get it.

```
- Path of `http-handler` and http `function` are exact match
- Path of `http-service` are matched with prefix
```

What I want to express in `matched with prefix` should be `matched with the longest prefix`, simple string match.

Reply:
> Would it be bad if we simply use longest match for everything?

My original thought is that both `http-handler` and http `function` are simple function, it's a limitation that they cannot do further routing inside the function.

Just chat with @carmenlau, she thinks it may be good to allow function also have a more flexible matching than exact match, because some 3rd party web hooks may put data in the path, e.g. user id or event name.

I have a pretty strong stand of keeping http `function` to be exact match because the function itself should not have any knowledge of the original request. But for `http-handler` since the function can receive original http request, so I think it's ok to make it have the same matching mechanism with `http-service`.

Reply:
agree.